### PR TITLE
ENG-684: add authenticated mux + descope integration

### PR DESCRIPTION
**Pull request by efiShtain**
# Overview

Added integration on client side for descope
Added support for authenticated access using auth mix
Added an example for authenticated call (with token) and unauthorized call

New env variables:
```
        "AK_AUTHENTICATOR__ENABLED": "true", // default false
        "AK_AUTHENTICATOR__PROVIDER": "descope", // currently only descope is supported
        "AK_AUTHENTICATOR__PROJECT_ID": "" // depends on provider, currently only project_id is required
```

## Details
This is client side authentication mostly, server only verifies the token but is not part of the authentication process it self
we can guard any endpoint we want

```Internal/backend/auth ``` has an authenticator interface that exposes a middleware that suppose to handle authentication. 

Default authentication is noAuthAuthentication that verifies nothing and sets a default user in context

Descpose authentication verifies the JWT and set basic user details in context

### CODE REVIEW

File: internal/backend/auth/auth.go
Comment (daabr) [inline internal/backend/auth/auth.go]: See lint error: define this key as a `struct{}` instead of a string to prevent collisions.

File: internal/backend/auth/auth.go
Comment (daabr) [inline internal/backend/auth/auth.go:53]: Why not a `default` case? Does Go complain that you're not returning anything?

File: internal/backend/auth/descope.go
Comment (daabr) [inline internal/backend/auth/descope.go]: Remove `": %w"`, the `zap.Error()` is a separate field in the log, not a part of the error message.

File: internal/backend/svc/svc.go
Comment (daabr) [inline internal/backend/svc/svc.go]: Question: if Descope is enabled, but I haven't logged in yet, or my login failed, won't this prevent me from seeing anything? Won't it affect images, CSS and scripts? Won't it affect integration webhooks?

**Review by daabr — COMMENTED** [review]
General comments:
1. Please fix various lint errors about unchecked errors
2. Embedded HTML files should be under `/web` (I was surprised about the main AK `index.html`, you don't have to fix that, but at least move your new files)
3. See my question in `svc.go` - I'll approve after you answer this
4. FYI but no action item needed - I don't like to combine CSS and JS inside HTML files, and using `/web` helps with this (see integrations), but it's not important to fix this in this PR or even soon

File: internal/backend/svc/svc.go
Comment (efiShtain) [reply internal/backend/svc/svc.go]: I might have put the authenticated route on the wrong place,
technically we would only want to block APIs and not web pages / standard http which is not grpc
I'll have a look

File: internal/backend/auth/auth.go
Comment (efiShtain) [reply internal/backend/auth/auth.go:53]: it doesn't really matter if you return from the cases

**Comment by efiShtain** [comment]
> General comments:
> 
> 1. Please fix various lint errors about unchecked errors
> 2. Embedded HTML files should be under `/web` (I was surprised about the main AK `index.html`, you don't have to fix that, but at least move your new files)
> 3. See my question in `svc.go` - I'll approve after you answer this
> 4. FYI but no action item needed - I don't like to combine CSS and JS inside HTML files, and using `/web` helps with this (see integrations), but it's not important to fix this in this PR or even soon

regarding 2, I'll move the files

regarding 4, everything here is under the assumption that this won't be the "real" UI, everything is only for dev purposes, so it doesn't really matter as long as it works
with that said, the direction the world goes with is everything on the same file (styling, backend logic, frontend logic, and more), look at any new frontend framework like nextjs / remix / svetle / vuejs and more. putting stuff together make it easier in lots of cases, especially sharing components. But again, it doesn't really matter for this case since it's not production UI, just dev stuff

File: internal/backend/auth/auth.go
Comment (efiShtain) [reply internal/backend/auth/auth.go]: fixed

File: internal/backend/auth/descope.go
Comment (efiShtain) [reply internal/backend/auth/descope.go]: good point, fixed

File: internal/backend/svc/svc.go
Comment (efiShtain) [reply internal/backend/svc/svc.go]: fixed it so only routes that begins with /api/ would be authenticated
next PR once this is approved is updating all our grpc endpoints to move to the authenticated routes

**Review by daabr — APPROVED** [review]

**Comment by itayd** [comment]
going to merge this and incorporate into my changes

**Pull request merged**
